esp-mbedtls/src/lib.rs

```diff
@@ -244,12 +244,133 @@ impl<'a> X509<'a> {
     }
 }
 
-/// Certificates used for a connection.
-///
-/// # Note:
-/// Both [certificate](Certificates::certificate) and [private_key](Certificates::private_key) must be set in pair.
-#[derive(Debug, Clone, Copy, PartialEq)]
-pub struct Certificates<'a> {
+/// Creates a wrapper over [mbedtls_x509_crt] to safely manage allocation and freeing on drop
+#[derive(Debug)]
+struct MbedTLSX509Crt(*mut mbedtls_x509_crt);
+
+impl MbedTLSX509Crt {
+
+    fn new(certificate: X509<'_>) -> Result<Self, TlsError> {
+        unsafe {
+            let ptr = aligned_calloc(
+                align_of::<mbedtls_x509_crt>(),
+                size_of::<mbedtls_x509_crt>(),
+            ) as *mut mbedtls_x509_crt;
+            if ptr.is_null() {
+                return Err(TlsError::OutOfMemory);
+            }
+            mbedtls_x509_crt_init(ptr);
+
+            let cleanup = || {
+                mbedtls_x509_crt_free(ptr);
+                free(ptr as *const _);
+            };
+
+            match certificate.format {
+                CertificateFormat::PEM => {
+                    error_checked!(
+                        mbedtls_x509_crt_parse(ptr, certificate.as_ptr(), certificate.len()),
+                        cleanup
+                    )?;
+                }
+                CertificateFormat::DER => {
+                    error_checked!(
+                        // TODO: This currently requires the X509 to live for the duration of this
+                        // lifetime
+                        mbedtls_x509_crt_parse_der_nocopy(
+                            ptr,
+                            certificate.as_ptr(),
+                            certificate.len()
+                        ),
+                        cleanup
+                    )?;
+                }
+            }
+            Ok(Self(ptr))
+        }
+    }
+}
+
+impl core::ops::Deref for MbedTLSX509Crt {
```

Comment:
Not sure we need this actually. Implementing `Deref` is always controversial (though it gives extra usability, but then - for an internal struct?...)